### PR TITLE
[ao][sparsity] Implementation of squash_mask()

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81814
* #80890
* __->__ #80889
* #80888
* #80887
* #80886

Unregisters aggreagate hook that was applied earlier and registers sparsification hooks.
The sparsification hook will apply the mask to the activations before it is fed into the
attached layer.

Test Plan:
```python test/test_ao_sparsity.py TestActivationSparsifier```

### CODE REVIEW

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 1]
```diff
@@ -270,10 +270,34 @@ def update_mask(self, name, data, configs):
...
+            else:
+                # apply per feature, feature_dim
+                for feature_idx in range(0, len(features)):
+                    feature = torch.Tensor([features[feature_idx]], device=input_data.device).long()
+                    sparsified = torch.index_select(input_data, feature_dim, feature) * mask[feature_idx]
+                    input_data.index_copy_(feature_dim, feature, sparsified)
+                return input_data
+        return hook
+
     def squash_mask(self, **kwargs):
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:294]: Aren't you supposed to deregister the hooks at this point? Does `.remove` do that?

Also, I think we should ask the user if they want to keep the forward pre hook. I am just thinking, if the user just want the masks, then the squash_mask is supposed to compute it, and get out, right? So a user can *optionally* request the forward hook to stay

**Review by z-a-f — CHANGES_REQUESTED** [review]

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:294]: yes ```.remove()``` deregisters the hook as mentioned in [docs](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_pre_hook.html)

Also, ```squash_mask()``` does not compute the mask, ```update_mask()``` does. We can add an argument ```attach_sparsify_hook``` which if True attache the sparsify forward pre-hook, otherwise just removes the aggregate handle

File: torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py [hunk 2]
```diff
@@ -270,10 +270,34 @@ def update_mask(self, name, data, configs):
                 data_feature = reduce_fn(data[feature_idx])
                 mask[feature_idx].data = mask_fn(data_feature, **sparse_config)
 
+    def __sparsify_hook(self, name):
```
Comment (jerryzh168) [inline torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py:273]: nit: rename to `sparsify_pre_hook` to make it clearer?

**Review by z-a-f — APPROVED** [review]
@macandro96 Check the tests + dynamo. Once fixed, you can merge
